Sources/Identity/CustomerInfo.swift

```diff
@@ -164,263 +153,168 @@ import Foundation
+    }
 
-        let subscriptionTransactionsByProductId: [String: [String: Any]]
-        let originalAppUserId: String
-        let managementURL: URL?
-        let originalApplicationVersion: String?
-        let originalPurchaseDate: Date?
-        let firstSeen: Date
-        let nonSubscriptionsByProductId: [String: [[String: Any]]]
-        let entitlementsData: [String: Any]
-        let nonSubscriptionTransactions: [StoreTransaction]
-        let allTransactionsByProductId: [String: [String: Any]]
-        let allPurchases: [String: [String: Any]]
-
-        init(subscriberData: [String: Any],
-             dateFormatter: DateFormatterType) throws {
-            let subscriptions = subscriberData["subscriptions"] as? [String: [String: Any]] ?? [:]
-            self.subscriptionTransactionsByProductId = subscriptions
-
-            // Metadata
-            self.originalApplicationVersion = subscriberData["original_application_version"] as? String
-
-            self.originalPurchaseDate =
-            dateFormatter.date(from: subscriberData["original_purchase_date"] as? String ?? "")
-
-            guard let firstSeenDateString = subscriberData["first_seen"] as? String else {
-                throw SubscriberDataError.firstSeenMissing
-            }
-
-            guard let firstSeenDate = dateFormatter.date(from: firstSeenDateString) else {
-                throw SubscriberDataError.firstSeenFormat
-            }
-
-            self.firstSeen = firstSeenDate
-
-            guard let originalAppUserIdString = subscriberData["original_app_user_id"] as? String else {
-                throw SubscriberDataError.originalAppUserIdMissing
-            }
-            self.originalAppUserId = originalAppUserIdString
-
-            self.managementURL = URL(string: subscriberData["management_url"] as? String ?? "")
-
-            // Purchases and entitlements
-            self.nonSubscriptionsByProductId =
-                subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [:]
-            self.entitlementsData = subscriberData["entitlements"] as? [String: Any] ?? [:]
-            self.nonSubscriptionTransactions = TransactionsFactory.nonSubscriptionTransactions(
-                withSubscriptionsData: nonSubscriptionsByProductId,
-                dateFormatter: dateFormatter)
-
-            let latestNonSubscriptionTransactionsByProductId = [String: [String: Any]](
-                uniqueKeysWithValues: nonSubscriptionsByProductId.map { productId, transactionsArray in
-                    (productId, transactionsArray.last ?? [:])
-                })
-
-            self.allTransactionsByProductId = latestNonSubscriptionTransactionsByProductId
-                .merging(subscriptionTransactionsByProductId, strategy: .keepOriginalValue)
-
-            self.allPurchases = latestNonSubscriptionTransactionsByProductId
-                .merging(subscriptionTransactionsByProductId) { (current, _) in current }
```

Comment:
This was super hacky and hard to follow. It was combining the dictionaries inside of "subscriptions" and "non_subscriptions", despite them having different formats. This is now made explicit through type conversions between the two.

Reply:
OMG, the new version is so ❤️ I was in here debugging something a while back and it was hard to follow. Great work!